docs/core/development/policy/compatibility-policy.rst

```diff
@@ -162,24 +162,38 @@ The most basic thing that can happen between Twisted versions, of course, is tha
 That means that no application may ever rely on, for example, the value of any **func_code** object's **co_code** attribute remaining stable, or the **checksum** of a .py file remaining stable.
 
 **Docstrings** may also change at any time.
-No application code may expect any Twisted class, module, or method's __doc__ attribute to remain the same.
+Applications must not depend on any Twisted class, module, or method's metadata attributes such as ``__module__``, ``__name__``, ``__qualname__``, ``__annotations__`` and ``__doc__`` to remain the same.
 
 
 New Attributes
 ^^^^^^^^^^^^^^
 
 New code may also be added.
-No application may ever rely on the output of the ``dir()`` function on any object remaining stable, nor on any object's ``__all__`` attribute, nor on any object's ``__dict__`` not having new keys added to it.
+Applications must not depend on the output of the ``dir()`` function on any object remaining stable, nor on any object's ``__all__`` attribute, nor on any object's ``__dict__`` not having new keys added to it.
 These may happen in any maintenance or bugfix release, no matter how minor.
 
 
 Pickling
 ^^^^^^^^
 
 Even though Python objects can be pickled and unpickled without explicit support for this, whether a particular pickled object can be unpickled after any particular change to the implementation of that object is less certain.
-Because of this, no application may depend on any object defined by Twisted to provide pickle compatibility between any release unless the object explicitly documents this as a feature it has.
+Because of this, applications must not depend on any object defined by Twisted to provide pickle compatibility between any release unless the object explicitly documents this as a feature it has.
 
 
+Representations
+^^^^^^^^^^^^^^^
+
+The printable representations of objects, as returned by ``repr(obj, /)`` and defined by ``def __repr__(self):`` are for debugging and informational purposes.
+Because of this, applications must not depend on any object defined by Twisted to provide repr compatibility between any release.
+descriptors, slots and weakrefs
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+How an object's attributes are defined and accessed is considered an implementation detail.
+To allow backwards compatibility an attribute may be moved from the instance ``__dict__`` into
+an ``@property`` or other descriptor based accessor.
+
+Assigning attributes to a constructed object, or monkey patching, is not considered a public use. This is to allow creating frozen and/or slotted classes. However if an object is weak referencable, an application may depend on it remaining weak referencable
+Because of this, applications must not depend on any object defined by Twisted to provide ``__dict__`` or ``__slots__`` compatibility between any release.
```

Comment:
The note about being weak referencable seems to be out of place here. Also I'm a bit worried about the implications of this: if a class is only weak referencable by accident and not by design, having that restriction on it could cause significant difficulties when making future changes to it.

Reply:
Yeah I'm not sure where to put it. Probably in the "Changes Covered by the Compatibility Policy".

>  if a class is only weak referencable by accident and not by design, having that restriction on it could cause significant difficulties when making future changes to it.

It's basically a footgun in `__slots__`, I don't think there's a good reason to make something not weak-referencable, and you can't do it in pypy anyway.

> Without a __weakref__ variable for each instance, classes defining __slots__ do not support weak references to its instances. If weak reference support is needed, then add '__weakref__' to the sequence of strings in the __slots__ declaration.

more: https://github.com/python-attrs/attrs/issues/174

Reply:
After discussion on IRC, it seems I misunderstood what it means to be weak referencable. So it might be worthwhile to define that term here, or link to a definition elsewhere if it already exists.

Reply:
If the reference to weak references (heh) is removed, the title here should lose it also.